### ISSUE TITLE
"Run All" from the dashboard is not working.

### ISSUE THREAD

**Issue by rupertlssmith** [mbuscemi/elm-test-runner#3]
I set up a project using 'elm-test init' as recommended.

Then I went to Packages -> Elm Test Runner -> Toggle Dashboard -> Run All, and I get:

    Compiler Error:
    No tests found in the test/ (or tests/) directory.

If I close the dashboard and to Packages -> Elm Test Runner -> Run All Tests, it works.
The Dashboard has to be closed though, if the dashboard is already open, I get the same error as above.

==== System details

    > uname -a
    Linux thinker 4.9.0-4-amd64 #1 SMP Debian 4.9.51-1 (2017-09-28) x86_64 GNU/Linux

    > elm-test --version
    0.18.10

    > atom --version
    Atom    : 1.21.1
    Electron: 1.6.15
    Chrome  : 56.0.2924.87
    Node    : 7.4.0

    elm-tests-runner: 0.1.1

**Comment by mbuscemi** [comment]
A couple of questions:
1. Does the root folder of your project workspace have your elm-package.json file? If not, please share your directory structure and where your elm project is compared to other files. I wrote some code to try to accommodate various setups, but this might not be perfect.
2. The run all dashboard button and cmd-t should use the same code. Do you get the same behavior with the command key shortcut, or different behavior?

**Comment by mbuscemi** [comment]
Also, the file you have currently open makes a difference. It would be good to know if the inconsistent behavior you're experiencing is truly related to the pane being opened/closed, or if opening a particular file has an effect. If the file comes from outside the Atom workspace with the Elm project, this could potentially be disruptive.

**Comment by rupertlssmith** [comment]
I have the elm-package.json in the root of the project. 
'elm-test init' created a second elm-package.json under tests/.

I have many projects open at once in Atom, in addition to this one that I am trying to test. I added them using "Add Project Folder".

CMD-T works, but only when I have my test file open.
If I close the test file, CMD-T or Run All Tests do not work.

I also now found the Run All from the dashboard can be made to work, I must click in the test file first to shift the focus onto it.

So basically, I think when the test file is the currently active file, things are working. When the test file is not the currently active file, it fails to locate the tests.

**Comment by mbuscemi** [comment]
Okay. I think I understand what is probably happening. Here's how the current setup looks for the place to run Elm tests:

1. Using the current open file, search all the open project root directories. If the current open file is underneath that directory, and if that directory has a elm-package.json, then use this directory, otherwise move on.

2. Starting with the directory that the current open file is in, walk "up" directories looking for an elm-package.json. If one is found, use that directory. Otherwise move on.

3. Walk all the project base directories looking for an elm-package.json. If you find one, stop, and assume this is your directory.

My current hypothesis is that you have multiple project folders open that have elm-package.json files in them, and step three is stopping on the first one it finds, though this is not the one with your tests. That would explain the current behavior. When you have a file open, it works because you're hitting either step 1 or 2. But 3 is not doing the right thing. How many of the projects in your workspace have an elm-package.json in the root directory?

**Comment by mbuscemi** [comment]
I've added some functionality to the 0.2.0 that I hope will correct this issue. I was able to reproduce the behavior you described by adding a project folder to my workspace that had an elm-package.json file in its root, but no tests. As anticipated, the test runner was picking it up as a "viable" directory and attempted to run tests against it, but none were there.

So, I've made a couple of changes to the flow I described above:

* For all steps, the code now ensures that there is an elm-package.json file present _and_ it ensures that there is a "tests" subdirectory. The steps will fail unless both conditions are met.
* I now check sub-directories of project directories to see if they meet the above criteria. This is essentially a new fourth step that happens if #3 above fails.

These changes will be released in 0.2.0.

**Comment by rupertlssmith** [comment]
Yes, I do have multiple Elm projects with elm-packages.json in them open at the same time.

Is it possible in Atom to find out the last file that was in focus? and try and use the project for that one?

**Comment by rupertlssmith** [comment]
My other suggestion is, once the Dashboard is opened, can you remember in its state what project it is the dashboard for?

**Comment by mbuscemi** [comment]
When I was doing my testing, I could make things work by making sure that the project with the tests was at the top of the list in the project sidebar. I wasn't sure if dragging and dropping projects around would work, but it did. That should get you working with the current version of the test runner.

It would be possible to track files opened and closed. Atom provides observer methods, which I suspect they're using for things like recent file/project history. I hesitate on these, though, because as much as possible, I want the behavior of the detection algorithm to be tied to the state of things "now" (from the end user's perspective). If I start detecting based on saved or cached state, then I would open up the possibility of UX problems that are even harder to deal with: users getting into funky states, and I would then need to ask them to pull their cached data and backtrack how the code got them into that state.

**Comment by mbuscemi** [comment]
0.2.0 has been released. @rupertlssmith, can you please confirm whether or not this is resolved?

**Comment by rupertlssmith** [comment]
The Run All button did work - but as I have multiple Elm projects open, it got the wrong one.

I opened a test file in a project lower down the list, and hit CMD-T, and it ran its tests. When I hit the Run All button though, it found a project higher up the list, and ran its tests instead.

**Comment by mbuscemi** [comment]
That is unsurprising, given there are now four steps in the determination process that respond to different states of the system. I have a proposal open (#4) to simplify the system in 0.3.0. Let me know if you have any feedback on it.

**Comment by mbuscemi** [comment]
FYI, after thinking through the design and user flow for this, I've come up with a plan that I think is both simpler from a code perspective and a better user experience, too. See the updated #4.

**Comment by mbuscemi** [comment]
@rupertlssmith 0.3.0 is released. Does this resolve your issue?
